### PR TITLE
Refactor DSL code

### PR DESCRIPTION
**Pull request by leszekhanusz**
Refactoring of the DSL code.

- The DSLSchema only needs a schema, a Client is no more needed.
- new `dsl_gql` function to convert the DSL operations into a Document
- this allows DSL queries to be executed in a session instead of a client (Solves #138)
- Added typing annotations
- Added code documentation and sphinx docs (Solves #82)
- supports Subscriptions
- fix nested input arguments
- allow to set the alias directly in the select method
- allow multiple operations in a document
- allow to set the operation name

Simple example:
```
query = dsl_gql(DSLQuery(ds.Query.hero.select(ds.Character.name)))                                                            
result = client.execute(query) 
```

### CODE REVIEW

**Comment by leszekhanusz** [comment]
I'll add documentation tomorrow before merging

File: gql/dsl.py [hunk 1]
```diff
@@ -155,10 +155,10 @@ def __init__(self, graphql_type: GraphQLTypeWithFields):
             Don't instanciate this class yourself.
             Use attributes of the :class:`DSLSchema` instead.
 
-        :param graphql_type: a GraphQL type
+        :param graphql_type: ther GraphQL type definition from the schema
```
Comment (KingDarBoja) [inline gql/dsl.py:158]: You made  typo here 👅

File: gql/dsl.py [hunk 2]
```diff
@@ -1,179 +1,362 @@
...
+def dsl_gql(*fields: "DSLField") -> DocumentNode:
+    """Given arguments of type :class:`DSLField` containing GraphQL requests,
+    generate a Document which can be executed later in a
+    gql client or a gql session.
+
+    Similar to the :func:`gql.gql` function but instead of parsing a python
+    string to describe the request, we are using requests which have been generated
+    dynamically using instances of :class:`DSLField` which have been generated
+    by instances of :class:`DSLType` which themselves have been generated from
+    a :class:`DSLSchema` class.
```
Comment (KingDarBoja) [inline gql/dsl.py:52]: I read too many `have been generated` phrases here, maybe something like:

> we are using requests which have been generated dynamically using instances of :class:`DSLField`, generated by instances of :class:`DSLType` which themselves originated from a :class:`DSLSchema` class.

File: gql/dsl.py [hunk 3]
```diff
@@ -1,179 +1,362 @@
...
+    """
+
+    # Check that we receive only arguments of type DSLField
+    # And that they are a root type
+    for field in fields:
+        if not isinstance(field, DSLField):
+            raise TypeError(
+                f"fields must be instances of DSLField. Received type: {type(field)}"
+            )
+        assert field.type_name in ["Query", "Mutation", "Subscription"], (
```
Comment (KingDarBoja) [inline gql/dsl.py:78]: Shouldn't this assertion match the enum types of the spec for [OperationType](https://spec.graphql.org/draft/#OperationType)?

Also graphql-core has [OperationType](https://github.com/graphql-python/graphql-core/blob/df8b5484a9e3ed69b1dcad66e923f9d1fea2cbdc/src/graphql/language/ast.py#L210) as an Enum.

**Review by KingDarBoja — COMMENTED** [review]
Outstanding work @leszekhanusz 

Also it seems you also fixed #104, right?

File: gql/dsl.py (hunk 1 above)
Comment (leszekhanusz) [reply gql/dsl.py]: Well spotted!

File: gql/dsl.py (hunk 2 above)
Comment (leszekhanusz) [reply gql/dsl.py]: Good idea

**Comment by leszekhanusz** [comment]
> Outstanding work @leszekhanusz

Thanks :smile: 

> Also it seems you also fixed #104, right?

No, this is not related, #104 is not using the DSL module at all.

File: gql/dsl.py [hunk 4]
```diff
@@ -1,179 +1,364 @@
...
+                          for this field.
+        """
+
         added_args = []
         for name, value in kwargs.items():
             arg = self.field.args.get(name)
             if not arg:
                 raise KeyError(f"Argument {name} does not exist in {self.field}.")
-            arg_type_serializer = get_arg_serializer(arg.type, known_serializers=dict())
+            arg_type_serializer = self._get_arg_serializer(arg.type)
```
Comment (KingDarBoja) [inline gql/dsl.py:310]: Unrelated but I always see this `_get_arg_serializer` requires `GraphQLInputType` but this type does not include `GraphQLInputField` type.

Seems like `graphql-core` always defines it as below:

```python
GraphQLInputType = Union[
    GraphQLScalarType, GraphQLEnumType, GraphQLInputObjectType, GraphQLWrappingType
]
```

@Cito this type definition seems to be based on https://spec.graphql.org/draft/#sec-Input-Objects.Type-Validation, right?

**Review by KingDarBoja — APPROVED** [review]
~Everything looks great, although I have made the alias change to work locally and tested it without any issue, should I push the commit to this branch so everything gets merged on one shot?~

Done 🍰

File: gql/dsl.py (hunk 4 above)
Comment (Cito) [reply gql/dsl.py:324]: GraphQLInputTypes are those that may be used as input types for arguments and directives. It's based on the same type in GraphQL.js which is based on the spec you linked.

File: gql/dsl.py [hunk 5]
```diff
@@ -1,179 +1,378 @@
...
+            inner_serializer = self._get_arg_serializer(arg_type.of_type)
+            return lambda list_values: ListValueNode(
+                values=FrozenList(inner_serializer(v) for v in list_values)
+            )
+        elif isinstance(arg_type, GraphQLEnumType):
+            return lambda value: EnumValueNode(
+                value=cast(GraphQLEnumType, arg_type).serialize(value)
+            )
 
+        return lambda value: ast_from_value(
```
Comment (Cito) [inline gql/dsl.py:362]: Maybe similar as above: `elif isinstance(arg_type, GraphQLScalarType): ... else: throw error`

File: gql/dsl.py [hunk 6]
```diff
@@ -1,179 +1,378 @@
...
 from graphql.pyutils import FrozenList
 
 from .utils import to_camel_case
 
+log = logging.getLogger(__name__)
 
-class DSLSchema(object):
-    def __init__(self, client):
-        self.client = client
+GraphQLTypeWithFields = Union[GraphQLObjectType, GraphQLInterfaceType]
```
Comment (Cito) [inline gql/dsl.py:39]: I would not create an alias for this, but use it inline. Also not that input types also have fields, so the name is misleading.

File: gql/dsl.py [hunk 7]
```diff
@@ -1,179 +1,378 @@
...
-        return self.execute(query(*args, **kwargs))
 
-    def mutate(self, *args, **kwargs):
-        return self.query(*args, operation="mutation", **kwargs)
+class DSLType:
+    """The DSLType represents a GraphQL type for the DSL code.
+
+    It can be a root type (Query, Mutation or Subscription).
+    Or it can be an interface type (Character in the StarWars schema).
+    Or it can be an object type (Human in the StarWars schema).
```
Comment (Cito) [inline gql/dsl.py:143]: The root types are also object types, so maybe rephrase like this:

It can be a root type (Query, Mutation or Subscription).
Or it can be any other object type (Human in the StarWars schema).
Or it can be an interface type (Character in the StarWars schema).

File: gql/dsl.py (hunk 5 above)
Comment (leszekhanusz) [reply gql/dsl.py:362]: The problem is that it is impossible in theory that the error happens if `_get_arg_serializer` is called by the `args` method.
So it means we have to add a meaningless test for this in order for the raise error line to get covered by the tests.
I propose instead to use `assert isinstance(arg_type, GraphQLScalarType)` here.

File: gql/dsl.py (hunk 6 above)
Comment (leszekhanusz) [reply gql/dsl.py]: Indeed, it is better. Done

File: gql/dsl.py (hunk 7 above)
Comment (leszekhanusz) [reply gql/dsl.py]: Done

File: gql/dsl.py (hunk 4 above)
Comment (leszekhanusz) [reply gql/dsl.py:322]: You are right @KingDarBoja , there was a problem here because the `_get_arg_serializer` method was called with a _GraphQLField_ argument instead of a type if the parent input type was a _GraphQLInputObjectType_.
I modified the code in my last commit to get the type directly instead of using the field so that now the typing info should be correct I think.

File: gql/dsl.py (hunk 5 above)
Comment (Cito) [reply gql/dsl.py:371]: Right, an assert may be appropriate in such a situation where you have a condition that should theoretically *never* fail. Note that it will be stripped away with "-O" which can be seen as an advantage (better performance in production) or disadvantage ("impossible" but still happening errors in production not detected). If you use asserts, then you should do it consistently for such cases in the whole codebase, (see also [here](https://github.com/graphql-python/graphql-core-legacy/issues/244)).

One of the reasons to add an "if isinstance" or an "assert" is that mypy should (in theory) understand that so that you don't need the "cast" (not sure if that works in the latest version of mypy). It is also a hint to the reader of the code. Btw, GraphQL-Core also has an `assert_scalar_type` function that you can use here. It raises a TypeError though instead of an AssertionError. Its return value is GraphQLScalarType so that should also help mypy.

I also add a "# pragma: no cover" to not decrease coverage for such code that should in theory never execute.

**Comment by leszekhanusz** [comment]
I can't believe it :man_facepalming: , the `_get_arg_serializer` function was not needed after all as all it did was to rewrite the `ast_from_value` method. It is now removed.

File: gql/dsl.py (hunk 5 above)
Comment (leszekhanusz) [reply gql/dsl.py]: See my last commit, this method was removed so this problem as well :laughing:

File: gql/dsl.py (hunk 4 above)
Comment (leszekhanusz) [reply gql/dsl.py]: all this is irrelevant now that the _get_arg_serializer method was removed :rofl:

**Review by KingDarBoja — APPROVED** [review]
Perfect, I loved the latest changes and the operation name feature :D
